### PR TITLE
Added support for assigning folder where ss_polipo.exe and polipo.conf are extracted

### PR DESCRIPTION
**Pull request by EraserKing**
This change aims to solve issues when system temp folder is not writable, or executing files from temp folder is blocked.

It added two command line arguments: /SameFolderForPolipo and /PolipoFolder.
If no arguments is used, ss_polipo.exe and polipo.conf are still extracted to temp folder as usual. User should not feel difference.
If /SameFolderForPolipo is used, these two files are extracted to the same folder where Shadowsocks.exe is located.
If /PolipoFolder <path> is used, these two files are extracted to the folder specified here.

### CODE REVIEW

**Comment by EraserKing** [comment]
Well, it seems AppVeyor ran into some trouble fetching source code...

File: shadowsocks-csharp/Controller/PolipoRunner.cs [hunk 1]
```diff
@@ -20,6 +20,32 @@ class PolipoRunner
         static PolipoRunner()
         {
             temppath = Path.GetTempPath();
+            string[] commandLineArgs = Environment.GetCommandLineArgs();
+            for (int i = 0; i < commandLineArgs.Length; i++)
+            {
+                string commandLineArg = commandLineArgs[i];
+                if (commandLineArg.ToLower() == "/SameFolderForPolipo".ToLower())
```
Comment (wenhx) [inline shadowsocks-csharp/Controller/PolipoRunner.cs]: You should use String.Equals(string a, string b, StringComparison comparisonType) <https://msdn.microsoft.com/en-us/library/t4411bks(v=vs.110).aspx>

File: shadowsocks-csharp/Controller/PolipoRunner.cs [hunk 2]
```diff
@@ -20,6 +20,32 @@ class PolipoRunner
...
             temppath = Path.GetTempPath();
+            string[] commandLineArgs = Environment.GetCommandLineArgs();
+            for (int i = 0; i < commandLineArgs.Length; i++)
+            {
+                string commandLineArg = commandLineArgs[i];
+                if (commandLineArg.ToLower() == "/SameFolderForPolipo".ToLower())
+                {
+                    temppath = Environment.CurrentDirectory;
+                }
+                if (commandLineArg.ToLower() == "/PolipoFolder".ToLower())
```
Comment (wenhx) [inline shadowsocks-csharp/Controller/PolipoRunner.cs]: Same as above.

File: shadowsocks-csharp/Controller/PolipoRunner.cs [hunk 3]
```diff
@@ -20,6 +20,32 @@ class PolipoRunner
...
+                {
+                    if (i == commandLineArgs.Length - 1)
+                    {
+                        System.Windows.Forms.MessageBox.Show("Polipo folder not specified.", "Error", System.Windows.Forms.MessageBoxButtons.OK, System.Windows.Forms.MessageBoxIcon.Error);
+                        throw new Exception("Polipo folder not specified");
+                    }
+                    else if (!IsDirectoryValid(commandLineArgs[i + 1]))
+                    {
+                        System.Windows.Forms.MessageBox.Show("Polipo folder not valid.", "Error", System.Windows.Forms.MessageBoxButtons.OK, System.Windows.Forms.MessageBoxIcon.Error);
+                        throw new Exception("Polipo folder not valid");
```
Comment (wenhx) [inline shadowsocks-csharp/Controller/PolipoRunner.cs]: Is here and the line 36 thrown exception will crash the process?

File: shadowsocks-csharp/Controller/PolipoRunner.cs [hunk 4]
```diff
@@ -122,5 +148,24 @@ private int GetFreePort()
             }
             throw new Exception("No free port found.");
         }
+
+        public static bool IsDirectoryValid(string path)
```
Comment (wenhx) [inline shadowsocks-csharp/Controller/PolipoRunner.cs]: This method does not appear to be called?

File: shadowsocks-csharp/Controller/PolipoRunner.cs (hunk 1 above)
Comment (EraserKing) [inline shadowsocks-csharp/Controller/PolipoRunner.cs]: All right. I'll open a new pr later.

File: shadowsocks-csharp/Controller/PolipoRunner.cs (hunk 3 above)
Comment (EraserKing) [inline shadowsocks-csharp/Controller/PolipoRunner.cs]: yes

File: shadowsocks-csharp/Controller/PolipoRunner.cs (hunk 4 above)
Comment (EraserKing) [inline shadowsocks-csharp/Controller/PolipoRunner.cs]: see line 38

**Comment by clowwindy** [comment]
I think this is not necessary since temp folder is not expected to be read only.

**Pull request closed without merging**